File: python3.8/cookiecutter-aws-sam-hello-python/{{cookiecutter.project_name}}/tests/integration/test_api_gateway.py
import os
from unittest import TestCase

import boto3
import requests


class TestApiGateway(TestCase):
    api_endpoint: str

    def setUp(self) -> None:
        """
        Based on the provided env variable AWS_SAM_STACK_NAME,
        here we use cloudformation API to find out what the HelloWorldApi URL is
        """
        stack_name = os.environ.get("AWS_SAM_STACK_NAME")
        self.failIf(not stack_name, "Cannot find env var AWS_SAM_STACK_NAME")

        client = boto3.client("cloudformation")

        response = client.describe_stacks(StackName=stack_name)
        stacks = response["Stacks"]
        self.assertTrue(stacks, f"Cannot find stack {stack_name}")

        stack_outputs = stacks[0]["Outputs"]
        api_outputs = [output for output in stack_outputs if output["OutputKey"] == "HelloWorldApi"]
        self.assertTrue(api_outputs, f"Cannot find output HelloWorldApi in stack {stack_name}")

        self.api_endpoint = api_outputs[0]["OutputValue"]

    def test_api_gateway(self):
        """
        Call the API Gateway endpoint and check the response
        """
        response = requests.get(self.api_endpoint)
        self.assertDictEqual(response.json(), {"message": "hello world"})

File: nodejs12.x/hello/{{cookiecutter.project_name}}/hello-world/tests/integration/test-api-gateway.js
"use strict";

const chai = require("chai");
const AWS = require("aws-sdk");
const https = require("https");
const expect = chai.expect;

/**
 * Get stack name from environment variable AWS_SAM_STACK_NAME.
 * throw exception if AWS_SAM_STACK_NAME is not set.
 */
const getStackName = () => {
  const stackName = process.env["AWS_SAM_STACK_NAME"];
  if (!stackName) {
    throw new Error(
      "Cannot find env var AWS_SAM_STACK_NAME.\n" +
        "Please setup this environment variable with the stack name where we are running integration tests."
    );
  }

  return stackName;
};

/**
 * Make sure env variable AWS_SAM_STACK_NAME exists with the name of the stack we are going to test.
 */
describe("Test API Gateway", function () {
  let apiEndpoint;

  /**
   * Based on the provided stack name,
   * here we use cloudformation API to find out what the HelloWorldApi URL is
   */
  before(async () => {
    const stackName = getStackName();

    const client = new AWS.CloudFormation();

    let response;
    try {
      response = await client
        .describeStacks({
          StackName: stackName,
        })
        .promise();
    } catch (e) {
      throw new Error(
        `Cannot find stack ${stackName}: ${e.message}\n` +
          `Please make sure stack with the name "${stackName}" exists.`
      );
    }

    const stacks = response.Stacks;

    const stackOutputs = stacks[0].Outputs;
    const apiOutput = stackOutputs.find(
      (output) => output.OutputKey === "HelloWorldApi"
    );

    expect(apiOutput, `Cannot find output HelloWorldApi in stack ${stackName}`)
      .not.to.be.undefined;

    apiEndpoint = apiOutput.OutputValue;
  });

  /**
   * Call the API Gateway endpoint and check the response
   */
  it("verifies successful response from api gateway", (done) => {
    console.info("api endpoint:", apiEndpoint);
    https
      .get(apiEndpoint, (res) => {
        expect(res.statusCode).to.be.equal(200);

        res.on("data", (data) => {
          const response = JSON.parse(data);
          expect(response).to.be.an("object");
          expect(response.message).to.be.equal("hello world");
          done();
        });
      })
      .on("error", (e) => {
        throw e;
      });
  });
});
